Update data entry navigation

lkleuver opened the pull request:
central storage of values and server responses for pollingstation form sections

controlled navigation through form sections
- go to the next section to fill in
- restrict navigation when the user forgot to save
- block sections that are later in the flow

[flowcharts describing the rendering and navigation logic](https://github.com/kiesraad/abacus/blob/main/documentatie/flowcharts/data-entry-forms.md)

# TODO while in draft:
- [x] fix current tests to check for navigation through the form
- [x] merge/rebase main
- [x] fix showing errors too early (differences form)
- [x] fix political group votes errors not showing at all
- [x] fix showing votes on candidates incorrect total error too early (voters and votes form) #149 
- [x] scroll up when there is an error on a page
- [x] improve aria-labels for icons
  - aria-labels should describe meaning of icon ("je bent hier") instead of appearance of icon ("pijl naar rechts")

# Testing findings 1

(by @jschuurk-kr)

## General
- [x] 1. The `console.log()`s  (e.g. "Checking changes!!!" and "new form state" with data) should be removed or be more informative → Fixed in 632aaf5ea51c6445a654e5d385cef31945f063d2
- [x] 2. Shift+Enter does not seem to work on any of the pages. (Ubuntu + Firefox) → Part of #263 
- [x] 3. Proposed url change: `/1/input/1/numbers` -> `/1/input/1/voters-and-votes` → Part of #184 
- [x] 4. Proposed url change: `:electionId/input` -> `elections/:electionId/input` → Part of #184
- [x] 5. Should we sort the errors and warnings, i.e. always show F201 above F202, etc? Currently they aren't. → Part of #266 
- [x] 6. If I have a warning and change the input so I get an error, the warning checkbox is very briefly rendered.
- [x] 7. Clicking "Volgende" seems to bring me to the furthest page visited, not necessarily the next page.

## Warnings checkbox
- [x] 1. The checkbox disappears even when not changing fields, e.g. when pressing ctrl, esc, alt, tab, arrow keys. → Will fix
- [x] 2. The checkbox disappears when replacing the value in a field with the value it contained. → Won't fix
- [x] 3. The checkbox often does not appear on the Differences page.
- [x] 4. If I uncheck the checkbox and click next, I don't get any feedback from the UI about what to do next.
  - There is [design for this in Figma](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=3395-73259&t=zRoBmvCy5Z9meMrV-4): an error "Je kan alleen verder als je het het papieren proces-verbaal hebt gecontroleerd." above the checkbox
- [x] 5. "Ik heb de aantallen gecontroleerd met papier en correct overgenomen." "met papier" sounds odd to me "met het papier"?
  - Relates to design, [added comment to Figma](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp?node-id=137-3939#919080233)
  - Copy is changed to "Ik heb de aantallen gecontroleerd met het papier en correct overgenomen." → Fixed in 26c0d3d2ac29e12de4abf7176cefa8ca45e79daf

## Voters and Votes page
- [x] 1. W209 (A-D == E-H) is shown twice, but shouldn't be, because it's a single error. → Fixed in fa3c5ad4489e71cf791e276295a3a39092d224d5

## Differences page
- [x] 1. "Te weinig uitgerekte stembiljetten" instead of "... uitgereikte ..." → Fixed in 8dbba0f846e2cb7d1117812dbf8c813ab1566391
- [x] 2. If I click the accept warnings checkbox and click "next", I'm not navigated to the next page.

## Political Party page
- [x] 1. Fill in, click next, does not navigate to next page.

## Abort/Save input popup (upon navigation)
- [x] 1. Change modal copy to match design, see https://github.com/kiesraad/abacus/pull/233#discussion_r1723618729.
  * Text in popup contains "TEMP: warnings" -> Fixed in fa3c5ad4489e71cf791e276295a3a39092d224d5
- [x] 2. How is a user to know what "melding sluiten" (X in the top right) does?
  * -> Close the modal without navigating away?
- [x] 3. Popup shows up too often. For example, go to Differences pages, go back to VotersAndVotes page, click on Differences nav item without making any changes, then the popup appears.
- [x] 4. If I click "Save input", I stay on the same page instead of going to the next one.
- [x] 5. If there's an error or warning on a page and I navigate to a different page without any changes to the input, the "Wat wil je doen met je invoer?" popup appears. There is no input to save or discard, so it doesn't make sense to show the popup.
- [x] 6. Popup does not match [design](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=3151-21014&t=L9Gz0ZZWlDQsHrgx-0).

## Abort/Save input popup ("Invoer afbreken" in top bar)
all issues related to this popup are to be addressed in https://github.com/kiesraad/abacus/issues/212
- [x] 1. If a page has an error or warning, both of the buttons of the "Invoer afbreken" modal have issues. Behavior is the same for errors and warnings:
  - "Invoer bewaren": first click nothing happens; second click POST request is successful, button is disabled, no navigation
  - "Niet bewaren": first click DELETE request is successful; second click navigates to page to select polling station

## F204
- [x] 1. After filling the party so that F204 is triggered, the icon disappears for "Aantal kiezers en stemmers". Unless that page had a warning, then the warning icon remains.
- [x] 2. To what degree has this been done? Both TODOs in issue #149 are still unchecked.

## Navigation
- [x] 1. I can get to the "Controleren en Opslaan" page by clicking on it on the navigation panel. Then two pages are shown as the active page. I need to change the url to get to an unbroken state. There's also a console error: `Uncaught (in promise) TypeError: form2 is null`.
- [x] 2. Submitting the last political group candidates votes page doesn't continue to the "Controleren en opslaan" page

## Navigation sidebar
- [x] 1. Pages in which all fields are empty have a green checkmark instead of a dash as icon.
- [x] 2. If I have an error on a page, then navigate to a different page ("niet bewaren"), the page in the nav bar has no icon. Should be an error icon. Or, navigation should not be possible and the user is shown a clear error message, they can't navigate away.

## Test  coverage
- [x] 1. React Testing Library tests for accepting warnings on the different forms. -> Joep
- [x] 2. React Testing Library tests for the warning you need to accept the warning(s) for the different forms. -> Joep
- [x] 3. Extend `PollingStation.test.tsx` to check icons in navigation panel. -> Les
- [x] 4. d2d test that triggers and corrects F204 error

praseodym commented:
How did you merge `main` into this branch? I don't see a merge commit but a bunch of commits from the `main` are listed here, which shouldn't happen. This makes the PR harder to review.

lkleuver commented:
git merge main
but it somehow switched to rebase

praseodym commented:
That's very strange, maybe you can get those commits out using a rebase and then try a regular merge?

praseodym commented:
Fixed together with @lkleuver, summary:

```shell
# remove up to commit before commits from main were pushed into this branch
git reset --hard 48a48edcb41204f0cee6fbb24bea11de64ca15fc
# merge main into this branch and fix merge conflicts
git fetch origin
git merge origin/main
# re-add commit made after commits from main branch
git cherry-pick 70b247ea48390bc101e4a46be494ea7e2fc71464
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.ts:
```diff
@@ -0,0 +1,135 @@
+
+export function isGlobalValidationResult(validationResult: ValidationResult): boolean {
+  switch (validationResult.code) {
+    case "F401":
```
F401 is not a global error, this is validated for each list when you press "Volgende". This just checks if the counts within the list add up to the given total.

lkleuver commented on frontend/app/component/form/differences/DifferencesForm.test.tsx:
```diff
@@ -202,40 +199,41 @@ describe("Test DifferencesForm", () => {
-      const feedbackServerError = await screen.findByTestId("feedback-server-error");
-      expect(feedbackServerError).toHaveTextContent(/^Error500 error from mock$/);
-    });
+    //Server errors are no longer rendered inside the form
+    // test("422 response results in display of error message", async () => {
+    //   overrideOnce("post", "/api/polling_stations/1/data_entries/1", 422, {
+    //     message: "422 error from mock",
+    //   });
+    //
+    //   const user = userEvent.setup();
+    //
+    //   renderForm();
+    //
+    //   const submitButton = await screen.findByRole("button", { name: "Volgende" });
+    //   await user.click(submitButton);
+    //   //TODO: server errors moved out of the form
+    //   //const feedbackServerError = await screen.findByTestId("feedback-server-error");
+    //   //expect(feedbackServerError).toHaveTextContent(/^Error422 error from mock$/);
+    // });
+
+    //Server errors are no longer rendered inside the form
+    // test("500 response results in display of error message", async () => {
+    //   overrideOnce("post", "/api/polling_stations/1/data_entries/1", 500, {
+    //     message: "500 error from mock",
+    //   });
+    //
+    //   const user = userEvent.setup();
+    //
+    //   renderForm();
+    //
+    //   const submitButton = await screen.findByRole("button", { name: "Volgende" });
+    //   await user.click(submitButton);
+    //   //TODO: server errors moved out of the form
+    //   //const feedbackServerError = await screen.findByTestId("feedback-server-error");
+    //   //expect(feedbackServerError).toHaveTextContent(/^Error500 error from mock$/);
+    // });
```
it's rendered on the same page but not by the form.
PollingStationFormNavigation renders it

praseodym replied:
I see! I'll clean up the tests.

jschuurk-kr replied:
Those commented out tests have an `OverrideOnce()` with a response that does not match the type `ErrorResponse`. That invalid response does match with what `PollingStationFormNavigation.tsx` expects: `{error.errorCode}: {error.message}`. (discovered this because I was looking into something else)

praseodym replied:
Fixed in 6b1435ec0ceb41fcb42609fef9d8b2fc2158a9e5

jschuurk-kr replied:
We need a better test for server errors, because if the backend returns a 422 with body `{"error": "some error text here"}, you get this in the frontend:

![image](https://github.com/user-attachments/assets/14e1f4c9-e0d9-4c86-95ef-ea7dee8bb14d)

(Also note that the styling seems off: error box looks a bit close to the heading.)

Lionqueen94 replied:
The server error is improved in https://github.com/kiesraad/abacus/pull/266 please check if you have any feedback on what exact text it should show there

praseodym replied:
> We need a better test for server errors, because if the backend returns a 422 with body `{"error": "some error text here"}, you get this in the frontend:
> 
> [..]

This should also be fixed in 6b1435ec0ceb41fcb42609fef9d8b2fc2158a9e5 because it parses the API POST response correctly.

jschuurk-kr commented:
# Testing findings 1

## General
1. The `console.log()`s  (e.g. "Checking changes!!!" and "new form state" with data) should be removed or be more informative
2. Shift+Enter does not seem to work on any of the pages. (Ubuntu + Firefox)
3. Proposed url change: `/1/input/1/numbers` -> `/1/input/1/voters-and-votes`
4. Proposed url change: `:electionId/input` -> `elections/:electionId/input`
5. Should we sort the errors and warnings, i.e. always show F201 above F202, etc? Currently they aren't.
6. If I have a warning and change the input so I get an error, the warning checkbox is very briefly rendered.
7. Clicking "Volgende" seems to bring me to the furthest page visited, not necessarily the next page.

## Warnings checkbox
1. The checkbox disappears even when not changing fields, e.g. when pressing ctrl, esc, alt, tab, arrow keys.
2. The checkbox disappears when replacing the value in a field with the value it contained.
3. The checkbox often does not appear on the Differences page.
4. If I uncheck the checkbox and click next, I don't get any feedback from the UI about what to do next.
5. "Ik heb de aantallen gecontroleerd met papier en correct overgenomen." "met papier" sounds odd to me "met het papier"?

## Voters and Votes page
1. W209 (A-D == E-H) is shown twice, but shouldn't be, because it's a single error.

## Differences page
1. "Te weinig uitgerekte stembiljetten" instead of "... uitgereikte ..."
2. If I click the accept warnings checkbox and click "next", I'm not navigated to the next page.

## Political Party page
1. Fill in, click next, does not navigate to next page.

## Abort/Save input popup
1. Text in popup contains "TEMP: warnings"
2. How is a user to know what "melding sluiten" (X in the top right) does?
3. Popup shows up too often. For example, go to Differences pages, go back to VotersAndVotes page, click on Differences nav item without making any changes, then the popup appears.
4. If I click "Save input", I stay on the same page instead of going to the next one.

## F204
1. After filling the party so that F204 is triggered, the icon disappears for "Aantal kiezers en stemmers". Unless that page had a warning, then the warning icon remains.
2. To what degree has this been done? Both TODOs in issue #149 are still unchecked.

## Navigation
1. I can get to the "Controleren en Opslaan" page by clicking on it on the navigation panel. Then two pages are shown as the active page. I need to change the url to get to an unbroken state. There's also a console error: `Uncaught (in promise) TypeError: form2 is null`.

Lionqueen94 commented:
@jschuurk-kr 
General.2 is on https://github.com/kiesraad/abacus/issues/101
General.3-4 will be done in https://github.com/kiesraad/abacus/issues/184 (elections/ is not listed there yet, but I agree with the request)
General.5 This is fixed in https://github.com/kiesraad/abacus/pull/266

jschuurk-kr commented:
> @jschuurk-kr General.2 is on #263 General.3-4 will be done in #184 (`elections/` is not listed there yet, but I agree with the request, might be good to specifically add the urls to be updated in that issue) General.5 This is fixed in #266

@Lionqueen94 Thanks! I updated my comment with the issue numbers. And added the proposed url changes to #184.

In the future, feel free to edit my comment to add issue numbers and/or other decisions.

jschuurk-kr commented on frontend/app/component/pollingstation/PollingStationProgress.tsx:
```diff
@@ -1,37 +1,76 @@
 
+  const { formState } = usePollingStationFormController();
+
+  const menuStatusForFormSection = React.useCallback(
+    (formSection?: FormSection): MenuStatus => {
+      if (!formSection) return "idle";
+      if (formSection.warnings.length > 0) {
+        return "warning";
+      }
+
+      if (formSection.id === formState.current) {
+        return "current";
+      }
+
+      if (formSection.errors.length === 0 && formSection.isSaved) {
+        return "accept";
+      }
+
+      return "idle";
+    },
+    [formState],
+  );
```
There are fewer states here than in `renderStatusIcon()` in `ProgressList.tsx`.

praseodym commented on frontend/lib/api/form/useErrorsAndWarnings.ts:
```diff
@@ -48,3 +53,49 @@ export function useErrorsAndWarnings(
   }, [errors, warnings, clientWarnings]);
 }
+
+//TODO: intermediate step for testing, might scrap code above.
```
What needs to happen to resolve this TODO?

lkleuver replied:
the useErrosandwarnings hook can be removed. then move the getErrorsandwarnings to a better place.

lkleuver replied:
refactored and added tests

jschuurk-kr commented on frontend/app/component/pollingstation/PollingStationFormNavigation.tsx:
```diff
@@ -0,0 +1,186 @@
+  return (
+    <>
+      {blocker.state === "blocked" && (
+        <Modal
+          onClose={() => {
+            blocker.proceed();
+          }}
+        >
+          <h2 id="modal-blocker-title">Wat wil je doen met je invoer?</h2>
+          <p>TEMP: {currentForm && reasonBlocked(formState, currentForm, values)}</p>
+          <p>
+            Ga je op een later moment verder met het invoeren van dit stembureau? Dan kan je de
+            invoer die je al hebt gedaan bewaren.
+            <br />
+            <br />
+            Twijfel je? Overleg dan met de coördinator.
+          </p>
+          <nav>
+            <Button
+              size="lg"
+              onClick={() => {
+                blocker.reset();
+                submitCurrentForm();
+              }}
+            >
+              Invoer bewaren
+            </Button>
+            <Button
+              size="lg"
+              variant="secondary"
+              onClick={() => {
+                blocker.proceed();
+              }}
+            >
+              Niet bewaren
+            </Button>
+          </nav>
+        </Modal>
```
We also have this modal in `AbortDataEntryControl.tsx`. Can we re-use that one here?

praseodym replied:
It shouldn't be the same modal, there is a different modal with different copy for this situation:

https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=3151-21014&t=NXsbLAd2Wx67Ma5j-4

jschuurk-kr commented on frontend/e2e-tests/dom-to-db-tests/data-entry.e2e.ts:
```diff
@@ -113,7 +123,114 @@ test.describe("errors and warnings", () => {
+
+    await expect(recountedPage.heading).toHaveText("Is er herteld?");
+    await expect(recountedPage.navRecounted).toHaveClass("active accept");
+    await expect(recountedPage.navVotersAndVotes).toHaveClass("idle current");
```
Test fails here, anticipating the removal of the distinction between `unsaved` and `current`.

jschuurk-kr replied:
commented out the failing `expect()`

Lionqueen94 commented on frontend/lib/api/form/pollingstation/PollingStationFormController.tsx:
```diff
@@ -151,12 +256,149 @@ export function PollingStationFormController({
+          }
+        }
+
+        //if the entire form is not completed yet, filter out global validation results since they dont have meaning yet.
```
```suggestion
        //if the entire form is not completed yet, filter out global validation results since they don't have meaning yet.
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/PollingStationFormController.tsx:
```diff
@@ -151,12 +256,149 @@ export function PollingStationFormController({
+        }
+        return newFormState;
+      });
+      //clean up
```
should there be clean-up here?

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+    formState.sections.voters_votes_counts.isSaved = true;
+    const validationResults: ValidationResult[] = [
+      {
+        fields: ["data.votes_counts.blank_votes_count"],
+        code: "F401",
```
```suggestion
        fields: ["data.differences_counts.fewer_ballots_count"],
        code: "F303",
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+
+    addValidationResultToFormState(formState, validationResults, "errors");
+
+    expect(formState.sections.voters_votes_counts.errors.length).toBe(1);
```
```suggestion
    expect(formState.sections.differences_counts.errors.length).toBe(1);
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+
+    const validationResults: ValidationResult[] = [
+      {
+        fields: ["data.votes_counts.invalid_votes_count", "data.political_group_votes[0]"],
+        code: "F401",
```
```suggestion
        fields: ["data.votes_counts.votes_candidates_counts", "data.political_group_votes"],
        code: "F204",
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+    addValidationResultToFormState(formState, validationResults, "errors");
+
+    expect(formState.sections.voters_votes_counts.errors.length).toBe(1);
+    expect(formState.sections.political_group_votes_1?.errors.length).toBe(1);
```
```suggestion
    expect(formState.sections.political_group_votes.errors.length).toBe(1);
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+
+    const validationResults: ValidationResult[] = [
+      {
+        fields: ["data.votes_counts.invalid_votes_count", "data.political_group_votes[0]"],
+        code: "F401",
```
```suggestion
        fields: ["data.votes_counts.votes_candidates_counts", "data.political_group_votes"],
        code: "F204",
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+    addValidationResultToFormState(formState, validationResults, "errors");
+
+    expect(formState.sections.voters_votes_counts.errors.length).toBe(0);
+    expect(formState.sections.political_group_votes_1?.errors.length).toBe(0);
```
```suggestion
    expect(formState.sections.political_group_votes.errors.length).toBe(0);
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+    expect(
+      isGlobalValidationResult({
+        code: "F204",
+        fields: ["data.votes_counts.blank_votes_count"],
```
```suggestion
        fields: ["data.votes_counts.votes_candidates_counts", "data.political_group_votes"],
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+    };
+    formState.sections.voters_votes_counts.errors = [
+      {
+        code: "W203",
```
```suggestion
        code: "W201",
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+
+    expect(
+      isGlobalValidationResult({
+        code: "W304",
```
```suggestion
        code: "W301",
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,403 @@
+
+    expect(
+      isGlobalValidationResult({
+        code: "F401",
+        fields: ["data.votes_counts.blank_votes_count"],
```
```suggestion
        code: "F303",
        fields: ["data.differences_counts.fewer_ballots_count"],
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/PollingStationFormController.tsx:
```diff
@@ -36,72 +30,183 @@ export interface PollingStationFormControllerProps {
+  defaultCurrentForm?: AnyFormReference | null;
+}
+
+//TODO: change getValues so it also works for political_group_votes, getValues should only return the form for the specific list to use in caching.
```
This TODO hasn't been solved yet right? Is it in scope for this PR?

jorisleker commented:
- [ ] nav-items that are disabled must not have a hover state, but they do
- [ ] "Controleren en opslaan" does not get a disabled state, even if it's not a link
- Design of "Controleren en opslaan" is not implemented yet, but I think that's out of scope?

lkleuver commented:
> * [ x]  nav-items that are disabled must not have a hover state, but they do
> * [ x]  "Controleren en opslaan" does not get a disabled state, even if it's not a link
> * [ x]  The "Je kan alleen verder als je het het papieren proces-verbaal hebt gecontroleerd." error feedback has `font-size: 1,125rem`, should be `1rem`;
> * [ ]  When the above error message is shown, the checkbox should have an error state as well (red border) (--> Ok to save styling of checkboxes to another issue though, not really related to implementing the scope of this ticket)

This is a tricky one, there is a todo in the Checkbox component, I have not found a way to style the border of the checkbox (without introducing a custom checkbox component).
Without a rounded border it should work but currently with, it's quite ugly.
A custom checkbox isn't a disaster btw but prefer to do that in a separate issue

> * Design of "Controleren en opslaan" is not implemented yet, but I think that's out of scope? -> [Epic: data entry check and finalisation #95](https://github.com/kiesraad/abacus/issues/95)

jschuurk-kr commented:
- [ ] Unchecking the checkbox for accepting warnings does not trigger the "Let op: niet opgeslagen wijzigingen" modal.
- [ ] If the furthest page has an error or warning, you make a change, and navigate away, the "Let op: niet opgeslagen wijzigingen" modal does not trigger.
- [ ] If you trigger the "Let op: niet opgeslagen wijzigingen" modal, close it by clicking the X, then abort input and click either of the buttons ("invoer bewaren" of "niet bewaren") , the "Let op: niet opgeslagen wijzigingen" modal is triggered again. If you click "invoer bewaren" you get into a broking state, where you can no longer abort input while saving your changes.
- [ ] I was unable to trigger the "fix or ignore" modal (see bottom of [this flowchart](https://github.com/kiesraad/abacus/blob/9684125a3ca87f4dc0f1bbc9d42a2ddf19c63296/documentatie/flowcharts/data-entry-forms.md#navigate-away-from-page)).

jschuurk-kr commented on frontend/app/module/input/PollingStation/PollingStation.test.tsx:
```diff
@@ -0,0 +1,325 @@
+const expectElementContainsIcon = (id: string, ariaLabel: string) => {
+  const el = screen.getByTestId(id);
+  expect(el).toBeInTheDocument();
+  expect(el).toContainHTML(`aria-label="${ariaLabel}"`);
```
beter: `expect(within(el).getByRole("img")).toHaveAccessibleName(ariaLabel);`

lkleuver replied:
would that still work if there are multiple images?

jschuurk-kr replied:
Problem is that `IconError` does not have a `role=img` attribute. Will commit and push a fix. Curiously, Playwright doesn't seem to need an explicit `role` and presumable infers it from  the `svg` tag (and the accessible name?).

It's not important to test the exact html structure. It is important for the test to get as close as possible to what a person would see. So asserting that the aria label is on an element with role image is better than asserting thatit's on any element. What we indeed should not do, is asserting the `aside` element is present.

jschuurk-kr commented on frontend/app/module/input/PollingStation/PollingStation.test.tsx:
```diff
@@ -0,0 +1,325 @@
+    await expectBlockerModal();
+  });
+
+  test("Progress list shows correct icons", async () => {
```
Am missing the icons "nog niet afgerond", "bevat een fout", "je bent hier".
Also see https://github.com/kiesraad/abacus/blob/133.frontend.update-data-entry-navigation/frontend/e2e-tests/dom-to-db-tests/data-entry.e2e.ts#L352

lkleuver replied:
added checks

jschuurk-kr commented:
Bug: clicking "Wijzigingen opslaan" after you correct an error and click an item in the navigation panel, navigates you to the next page, instead of the page you clicked on in the navigation panel.

Steps to reproduce:
1. Fill the Recounted and Differences page with valid input. (so no errors, no warnings).
2. Return to the Differences page.
3. Fill in the Differences page so that it results in an error, and click "Next".
4. Fix the input so that it should no longer generate an error. Do __not__ click "Next".
5. Click "Is er herteld?" in the navigation panel.
6. Click "Wijzigingen opslaan" in the modal.

Result: user is navigated to the Differences page instead of the Recounted page.

praseodym commented on frontend/app/component/form/candidates_votes_form/CandidatesVotesForm.tsx:
```diff
@@ -31,79 +39,93 @@ export interface CandidatesVotesFormProps {
+  }, [deformat, group]);
+
+  const getIgnoreWarnings = React.useCallback(() => {
+    const checkbox = document.getElementById(_IGNORE_WARNINGS_ID) as HTMLInputElement | null;
```
Why not use a React `ref` for this? (And same for the checkbox in other form components)

lkleuver replied:
this is one of things I would like to change, store all values and changes directly in a ref type object centrally. 
reduces complexity and test timing failures.

in this case, ref was a bit flaky when called from the PollingStationNavigation

praseodym replied:
Maybe because the value can be null, but `getElementById` has the same issue.. so not sure why a ref is more flaky than this. We can look at it later though.

praseodym replied:
I will create an issue for this.

praseodym replied:
#290

praseodym commented on frontend/lib/ui/BottomBar/BottomBar.tsx:
```diff
@@ -12,3 +12,13 @@ export interface BottomBarProps {
+  children: React.ReactNode;
+  hidden?: boolean;
+}) {
+  return <section className={cn("row", { hidden: !!hidden })}>{children}</section>;
```
```suggestion
  return <section className={cn("row", { hidden: hidden })}>{children}</section>;
```

lkleuver commented:
> Bug: clicking "Wijzigingen opslaan" after you correct an error and click an item in the navigation panel, navigates you to the next page, instead of the page you clicked on in the navigation panel.
> 
> Steps to reproduce:
> 
> 1. Fill the Recounted and Differences page with valid input. (so no errors, no warnings).
> 2. Return to the Differences page.
> 3. Fill in the Differences page so that it results in an error, and click "Next".
> 4. Fix the input so that it should no longer generate an error. Do **not** click "Next".
> 5. Click "Is er herteld?" in the navigation panel.
> 6. Click "Wijzigingen opslaan" in the modal.
> 
> Result: user is navigated to the Differences page instead of the Recounted page.

Fixed bug and added test

jorisleker commented:
Found a small bug. Reproduction steps 
- test on a viewport that is lower then the height of the form, so you need to scroll down to reach the 'volgende' button.
- on `/numbers` enter the following: A: 30, D: 30, E:10, F:10, G:10, H30 
- click 'volgende', this will trigger two warnings (W201 + W202) and scroll you back up
- the 'accept warnings' checkbox is shown (correct). Don't check it
- change the input F:9, G:11
- the 'accept warnings' checkbox is hidden (correct)
- click 'volgende'
- this triggers two warnings (W201 + W202)

observed behavior:
- warnings are shown (correct)
- checkbox 'accept warnings' is shown (correct)
- page does not scroll up (incorrect)

expected behavior:
- warnings are shown
- checkbox 'accept warnings' is shown
- page scrolls up

jschuurk-kr commented:
Bug: going back to Recounted page and changing recounted from "no" to "yes" leads to behavior that's confusing to the user.

Steps to reproduce:
1. Fill in the forms with recounted set to "no" until you reach "Controleren en opslaan".
2. Go back to the Recounted page, set recounted to "yes", click "Volgende".
3. This leads you to the Voters and Votes page.
4. Click in the navigation panel on "Controleren en opslaan".
5. Unsaved changes modal appears.
6. Click "niet opslaan" in the unsaved changes modal.
7. You're now on the Controleren en Opslaan-page, without any errors or warnings.

Issues:
- The unsaved changes modal in step 5 appears for a page on which the user did not make any changes.
- Recounted was set to "yes", no values where filled in for the four recounted fields, there are no errors or warnings. This is an inconsistent state.

@lkleuver is looking into how we can fix this. @praseodym would be good to have your input as well.

jschuurk-kr commented on frontend/app/component/pollingstation/PollingStationFormNavigation.tsx:
```diff
@@ -0,0 +1,206 @@
+    if (formSection.warnings.length > 0 && !formSection.ignoreWarnings) {
+      result.push("warnings");
+    }
+    //if (!formSection.isSubmitted) {
```
This line (and line 202) can be removed?

lkleuver replied:
removed

jschuurk-kr commented on frontend/app/component/form/candidates_votes_form/CandidatesVotesForm.test.tsx:
```diff
@@ -1,5 +1,22 @@
+/** 
+ * Test Plan
```
I'd prefer to remove the Test Plan comments in favor of (1) moving the use case document with errors and warnings here; (2) making that document easier to read / get an overview from.

lkleuver replied:
removed

lkleuver commented:
> Bug: going back to Recounted page and changing recounted from "no" to "yes" leads to behavior that's confusing to the user.
> 
> Steps to reproduce:
> 
> 1. Fill in the forms with recounted set to "no" until you reach "Controleren en opslaan".
> 2. Go back to the Recounted page, set recounted to "yes", click "Volgende".
> 3. This leads you to the Voters and Votes page.
> 4. Click in the navigation panel on "Controleren en opslaan".
> 5. Unsaved changes modal appears.
> 6. Click "niet opslaan" in the unsaved changes modal.
> 7. You're now on the Controleren en Opslaan-page, without any errors or warnings.
> 
> Issues:
> 
> * The unsaved changes modal in step 5 appears for a page on which the user did not make any changes.
> * Recounted was set to "yes", no values where filled in for the four recounted fields, there are no errors or warnings. This is an inconsistent state.
> 
> @lkleuver is looking into how we can fix this. @praseodym would be good to have your input as well.

I've tested with setting voters_recounts values to 0 if recounted was previous no and changed to yes.
This stops the "has changes modal" for voters_and_votes to show and shows an error icon for the differences form.

We need to determine what the expected/desired behavior should be.

jschuurk-kr commented:
> I've tested with setting voters_recounts values to 0 if recounted was previous no and changed to yes. This stops the "has changes modal" for voters_and_votes to show and shows an error icon for the differences form.
> 
> We need to determine what the expected/desired behavior should be.

Discussed this and implemented fix is desired behavior. So next step is adding an integration test for the behavior.

Lionqueen94 commented on frontend/e2e-tests/dom-to-db-tests/data-entry.e2e.ts:
```diff
@@ -85,16 +90,88 @@ test.describe("errors and warnings", () => {
+
+    await votersVotesPage.navPanel.list(1).click();
+    await expect(candidatesListPage_1.heading).toBeVisible();
+    // fill counts of List 1 with data that doet match the total votes on candidates
```
```suggestion
    // fill counts of List 1 with data that does match the total votes on candidates
```

Lionqueen94 commented on frontend/e2e-tests/dom-to-db-tests/data-entry.e2e.ts:
```diff
@@ -85,16 +90,88 @@ test.describe("errors and warnings", () => {
+    await differencesPage.next.click();
+
+    const candidatesListPage_1 = new CandidatesListPage(page, "Lijst 1 - Political Group A");
+    // fill counts of List 1 with data that doet not match the total votes on candidates
```
```suggestion
    // fill counts of List 1 with data that does not match the total votes on candidates
```

Lionqueen94 commented on frontend/lib/api/form/pollingstation/PollingStationFormController.tsx:
```diff
@@ -151,12 +277,171 @@ export function PollingStationFormController({
+        const newFormState = { ...old };
+        //reset all errors/warnings, and submitted, the server validates the entire request each time.
+        //a reset is done before submitting the form to the server.
+        //resetFormSectionState(newFormState);
```
```suggestion
```

Commented out lines should be removed

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,417 @@
+    expect(formState.sections.differences_counts.errors.length).toBe(1);
+  });
+
+  // FIXME: F.204 is a global error, so is this test relevant?
```
This should be addressed.

lkleuver replied:
this is a test technically for a global error (or specifically errors that span multiple sections)

Lionqueen94 replied:
Ok then it's best to remove the FIXME comment :+1:

lkleuver replied:
I removed it... right, at least I cant find it 😅 ? also slightly tweaked the test

Lionqueen94 commented on frontend/lib/api/form/pollingstation/pollingStationUtils.test.ts:
```diff
@@ -0,0 +1,417 @@
+    const errorsAndWarnings = getErrorsAndWarnings(errors, warnings, clientWarnings);
+    expect(errorsAndWarnings.get("blank_votes_count")).toBeDefined();
+    expect(errorsAndWarnings.get("blank_votes_count")?.errors.length).toBe(1);
+    //warnings should not be added if errors.
+    expect(errorsAndWarnings.get("blank_votes_count")?.warnings.length).toBe(1);
```
The comment states that warnings should not be added if there are errors but the length of the warnings is 1, so they are added?

jschuurk-kr replied:
The warning that's present is from the `clientWarnings` defined in the test. We're missing a test with errors, warnings, and no client warnings.

Or we should validate on something more specific than the length of an array, which might be a good idea in any case.

lkleuver replied:
I made the test more specific.
Also the clientWarnings will hopefully be removed when moving logic to Input component.
